Framework/Core/include/Framework/PartRef.h

```diff
@@ -21,9 +21,66 @@ namespace framework
 {
 
 /// Reference to an inflight part.
-struct PartRef {
-  std::unique_ptr<FairMQMessage> header;
-  std::unique_ptr<FairMQMessage> payload;
+class PartRef
+{
+ public:
+  PartRef()
+    : mMessages(2)
+  {
+  }
+  PartRef(FairMQMessagePtr&& h, FairMQMessagePtr&& p)
+    : mMessages()
+  {
+    mMessages.emplace_back(std::move(h));
+    mMessages.emplace_back(std::move(p));
+  }
+
+  PartRef(std::vector<FairMQMessagePtr>&& m)
+    : mMessages(std::move(m))
+  {
+  }
+
+  PartRef(PartRef&& ref)
+    : mMessages(std::move(ref.mMessages))
+  {
+  }
+
+  PartRef& operator=(PartRef&& ref)
+  {
+    if (&ref != this) {
+      mMessages = std::move(ref.mMessages);
+    }
+    return *this;
+  }
+
+  std::unique_ptr<FairMQMessage>& header()
+  {
+    return mMessages[0];
+  }
+
+  std::unique_ptr<FairMQMessage> const& header() const
+  {
+    return mMessages[0];
+  }
+
+  std::unique_ptr<FairMQMessage>& payload(size_t part = 0)
+  {
+    return mMessages[part + 1];
+  }
+
+  std::unique_ptr<FairMQMessage> const& payload(size_t part = 0) const
+  {
+    return mMessages[part + 1];
+  }
+
+  size_t getNumberOfPayloads() const
+  {
+    // we always have 2 message slots
+    return mMessages.size() - 1;
+  }
+
+ private:
+  std::vector<std::unique_ptr<FairMQMessage>> mMessages;
```

Comment:
I am a bit worried we will start getting vector allocations all over the place even for the (common?) case of a single {header,payload} pair. Is there any way we can avoid that?

Reply:
sharing your concern, we could have the vector in addition to the two standard members and handle it in the access functions. Assuming that an empty vector has less impact in terms of allocation. To have some numbers I will do a benchmark.

Reply:
Yes, sounds like a good idea, waiting for the benchmark.

Reply:
This has been fixed, PartRef is deprecated now, MessageSet is using a linear storage for all messages belonging to one cell in the DataRelayer cache